packages/sync-service/lib/electric/plug/serve_shape_plug.ex

```diff
@@ -267,9 +268,13 @@ defmodule Electric.Plug.ServeShapePlug do
       {:ok, {offset, snapshot}} ->
         log =
           Shapes.get_log_stream(conn.assigns.config, shape_id, since: offset, up_to: last_offset)
-          |> Enum.to_list()
 
-        send_resp(conn, 200, Jason.encode_to_iodata!(snapshot ++ log ++ @up_to_date))
+        snapshot
+        |> Stream.concat(log)
+        |> Stream.concat(@up_to_date)
+        |> to_json_stream()
+        |> Stream.chunk_every(500)
```

Comment:
This is great!

This can be further improved by switching the last two steps, though we'll need to change the way the client decodes the response too.

In your implementation, every entry in the concatenated stream is encoded to JSON individually. But we could instead perform the chunking first, then encode each chunk to JSON array. This would make the shape of the response payload different, an array of arrays, but if the client knows to expect that, I'm expecting even better performance from the reduce number of `Jason.encode` calls.

Reply:
Another reason to want to chunk the stream _before_ the JSON encoding pass is to optimize how the client consumes the stream. Currently our client accumulates the complete array in memory before processing it. I'm not sure if it's possible to parse a JSON array in a streaming fashion in JS, but it's definitely possible to use line-delimited JSON instead of wrapping the stream of chunks in a top-level JSON array. Then the client would be able to process messages one chunk at a time instead of waiting for the closing bracket of the top-level array.

On the server side, it might be possible to use the new `json` module introduced in Erlang/OTP 27.0 to adjust encoding of a list into a JSON array to omit the surrounding brackets. Then we would be able to encode the whole chunk of the stream while maintaining the same shape if a single JSON array for the overall response payload.

Reply:
BTW, this is a nitpick, but I find this version with fewer `Stream.concat()` calls more readable:
```suggestion
        [snapshot, log, @up_to_date]
        |> Stream.concat()
        |> to_json_stream()
        |> Stream.chunk_every(500)
```

Reply:
we should sanity-check that there is an appreciable difference between concatenating individual rows or the chunk.

It wouldn't be great to change the decoding on the client because of this, specially knowing that some libs can already take out the brackets. But given that we do the encoding, wouldn't it be trivial to remove the first and last bytes when storing to file?

In any case, It sounds good to use the new json lib also because it seems faster than jason.

Reply:
re) json lines — I did experiment with that early on while prototyping and at least the library I tried is extreamly slow:

> I was curious to get a ballpark figure on how long it'd take to load data via jsonlines over http — so I generated files with 1mb and 10mb of fake user data — 1mb takes around 400-450ms to parse on my mac and 10mb takes ~3.5mb — the time should be dominated by parsing not network time so these are roughly correct numbers for loading over the public internet.
> 
> I'm using a random jsonl parsing library off npm — so potentially that can be sped up too

Browser JSON parsers are heavily optimized so userland implementations seem to suffer speed-wise by comparison.

I do agree that avoiding memory spikes when parsing very large JSON arrays is a great idea. But what would also work is is restricting how much data we send at a time to the user. We can do that while responding to requests. E.g. the initial request returns offsets 0-100,000 and then the next request returns offsets 100,001-178,340.

We'll also want to do this as CDNs balk at too large of response payloads e.g. Fastly gets grumpy if payloads are larger than 20mb.

Reply:
Yeah. We need to put limits on the maximum amount of data that goes out on each request so everything is more predictable.
I think is just a little step from where we are now.
Shall we look at it as a reliability issue and not a premature optimization ? :) 

There is a small complexity with initial chunk of changes not being “addressable”, but solvable.

Reply:
Great discussion!

This PR looks good to merge to me. We can follow it up with further investigation of chunking-related improvements.